packages/dates/src/format.ts

```diff
@@ -0,0 +1,166 @@
+  | 'm'
+  | 'ss'
+  | 's'
+  | 'A'
+  | 'a'
+  | 'dddd'
+  | 'ddd';
+
+type TokenDateTimeFormatOptionsMap = {
+  [key in Token]: {
+    type: Intl.DateTimeFormatPartTypes;
+    options: DateTimeFormatOptions;
+    extraParser?: (str: string) => string;
+  };
+};
+
+const TOKEN_MAP: TokenDateTimeFormatOptionsMap = {
+  YYYY: {type: 'year', options: {year: 'numeric'}},
+  YY: {type: 'year', options: {year: '2-digit'}},
+  MMMM: {type: 'month', options: {month: 'long'}},
+  MMM: {type: 'month', options: {month: 'short'}},
+  MM: {type: 'month', options: {month: '2-digit'}},
+  // eslint-disable-next-line id-length
+  M: {type: 'month', options: {month: 'numeric'}},
+  DD: {type: 'day', options: {day: '2-digit'}},
+  // eslint-disable-next-line id-length
+  D: {type: 'day', options: {day: 'numeric'}},
+  HH: {type: 'hour', options: {hour: '2-digit', hour12: false}},
+  // eslint-disable-next-line id-length
+  H: {
+    type: 'hour',
+    options: {hour: 'numeric', hour12: false},
+    // extraParser is needed when Node and Browsers don't respect the 2-digit
+    // and numeric options, which happens occasionally for time units
+    extraParser: numericFromTwoDigit,
+  },
+  hh: {type: 'hour', options: {hour: '2-digit', hour12: true}},
+  // eslint-disable-next-line id-length
+  h: {type: 'hour', options: {hour: 'numeric', hour12: true}},
+  mm: {
+    type: 'minute',
+    options: {minute: '2-digit'},
+    // see 'H' token parsing for more details about extraParser
+    extraParser: twoDigitFromNumeric,
+  },
+  // eslint-disable-next-line id-length
+  m: {type: 'minute', options: {minute: 'numeric'}},
+  ss: {
+    type: 'second',
+    options: {second: '2-digit'},
+    extraParser: twoDigitFromNumeric,
+  },
+  // eslint-disable-next-line id-length
+  s: {type: 'second', options: {second: 'numeric'}},
+  // eslint-disable-next-line id-length
+  A: {
+    type: 'dayPeriod',
+    options: {
+      hour: '2-digit',
+      minute: '2-digit',
+      second: '2-digit',
+      hour12: true,
+    },
+  },
+  // eslint-disable-next-line id-length
+  a: {
+    type: 'dayPeriod',
+    options: {
+      hour: '2-digit',
+      minute: '2-digit',
+      second: '2-digit',
+      hour12: true,
+    },
+    extraParser: lowercase,
+  },
+  dddd: {type: 'weekday', options: {weekday: 'long'}},
+  ddd: {type: 'weekday', options: {weekday: 'short'}},
+};
+
+function numericFromTwoDigit(time: string) {
+  return time.startsWith('0') && time.length === 2 ? time.charAt(1) : time;
+}
+
+function twoDigitFromNumeric(time: string) {
+  return time.length === 1 ? `0${time}` : time;
+}
+
+function lowercase(time: string) {
+  return time.toLowerCase();
+}
+
+export function format(
+  date: Date,
+  fmtStr: string,
+  tz?: string,
+  localeStr?: string,
+): string {
+  const {locale, timeZone} = memoizedGetDateTimeFormat(localeStr, {
+    timeZone: tz,
+  }).resolvedOptions();
+  const allTokens = new RegExp(
+    Object.keys(TOKEN_MAP)
+      .map((ch) => `(${ch})`)
```

Comment:
You probably want `^(FOO|BAR|BAZ)$` instead of `(FOO)|(BAR)|(BAZ)`

1. Move the capture parens outside
2. Anchor the start and end -- you don't want to accidentally capture other crap

Reply:
Nope. As we saw with the failing tests, we shouldn't anchor :)

Reply:
But should still `/(FOO|BAR|BAZ)/` to reduce the number of match tokens